### PR TITLE
Add com.github.micahflee.torbrowser-launcher

### PR DESCRIPTION
**Pull request by AsciiWolf**
Better version of #1066 based on [Tor Browser Launcher](https://github.com/micahflee/torbrowser-launcher). The latest Tor Browser is automatically downloaded and unpacked by the launcher and everything works the same as when using regular Tor Browser Launcher from the system repositories (it is packaged in Fedora, Ubuntu and other distributions). No Tor Browser profile in `/tmp` and other ugly hacks and no need to update the Flatpak every time a new Tor Browser tarball is released - Tor Browser Launcher handles this itself in a secure way (gpg signature verification). I have been using this Flatpak-ed version for a few hours and everything seems to work absolutely fine. :-)

### CODE REVIEW

**Comment by AsciiWolf** [comment]
bot, build com.github.micahflee.torbrowser-launcher

**Comment by AsciiWolf** [comment]
bot, build com.github.micahflee.torbrowser-launcher

**Comment by AsciiWolf** [comment]
bot, build com.github.micahflee.torbrowser-launcher

File: com.github.micahflee.torbrowser-launcher.yaml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+          url: https://files.pythonhosted.org/packages/fc/bb/a5768c230f9ddb03acc9ef3f0d4a3cf93462473795d18e9535498c8f929d/chardet-3.0.4.tar.gz
+          sha256: 84ab92ed1c4d4f16916e05906b6b75a6c0fb5db821cc65e70cbd64a3e2a5eaae
+
+  - name: torbrowser-launcher
+    buildsystem: simple
+    sources:
+        - type: archive
+          url: https://github.com/micahflee/torbrowser-launcher/archive/v0.3.2.tar.gz
+          sha256: 25ad4b0c70579628bf18f4cdcd18ac303fe4f7f1c590cc48df2a87651c92991c
+        - type: patch
```
Comment (gasinvein) [inline com.github.micahflee.torbrowser-launcher.yaml]: You can use single `patch` source with multiple paths, e.g.
```yaml
- type: patch
  paths:
    - remove-absolute-paths.patch
    - use-standard-icon-location.patch
    - ...
```

File: com.github.micahflee.torbrowser-launcher.yaml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+rename-appdata-file: torbrowser.appdata.xml
+rename-icon: torbrowser
+finish-args:
+  - --share=ipc
+  - --share=network
+  - --socket=x11
+  - --device=dri
+  - --socket=pulseaudio
+  - --persist=.config/torbrowser
+  - --persist=.local/share/torbrowser
```
Comment (gasinvein) [inline com.github.micahflee.torbrowser-launcher.yaml]: Is it hardcoded in the launcher or in TorBrowser itself? Is the former, maybe it's better to patch this, too?

File: com.github.micahflee.torbrowser-launcher.yaml (hunk 1 above)
Comment (AsciiWolf) [reply com.github.micahflee.torbrowser-launcher.yaml]: Nice, I didn't know that this is possible. Thanks!

File: com.github.micahflee.torbrowser-launcher.yaml (hunk 2 above)
Comment (AsciiWolf) [reply com.github.micahflee.torbrowser-launcher.yaml]: Should be in the launcher. However, I currently don't have enough time to look into the source code. Also, I didn't want to make any unnecessary modifications.

**Comment by AsciiWolf** [comment]
@gasinvein By the way, are you interested in push rights to the repo after it is on Flathub? I will be more than happy if someone helps with maintaining this. :-)

File: com.github.micahflee.torbrowser-launcher.yaml (hunk 1 above)
Comment (gasinvein) [reply com.github.micahflee.torbrowser-launcher.yaml]: It actually wasn't until [recently](https://github.com/flatpak/flatpak-builder/releases/tag/1.0.0).

File: com.github.micahflee.torbrowser-launcher.yaml (hunk 2 above)
Comment (gasinvein) [reply com.github.micahflee.torbrowser-launcher.yaml]: I'd rather consider such changes at least highly desirable. Will look into it later.

**Comment by gasinvein** [comment]
@AsciiWolf I don't use Tor Browser and don't really want to maintain it, honestly, but feel free to CC me anyway.

**Comment by AsciiWolf** [comment]
bot, build com.github.micahflee.torbrowser-launcher

File: com.github.micahflee.torbrowser-launcher.yaml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  - --filesystem=xdg-download
+modules:
+  - shared-modules/dbus-glib/dbus-glib-0.110.json
+
+  - name: python3-sip
+    buildsystem: simple
+    build-commands:
+      - python3 configure.py
```
Comment (gasinvein) [inline com.github.micahflee.torbrowser-launcher.yaml]: Hm, I'd rather rewrite it as a native autotools module, e.g. add a `shell` source with command `ln -s configure.py configure` and use `config-args`. But it's up to you.

**Comment by AsciiWolf** [comment]
By the way, upstream was [contacted](https://github.com/micahflee/torbrowser-launcher/issues/407#issuecomment-526147710) and relevant patches were sent as PRs.

File: com.github.micahflee.torbrowser-launcher.yaml (hunk 2 above)
Comment (tilosp) [reply com.github.micahflee.torbrowser-launcher.yaml]: I opened a pr (https://github.com/AsciiWolf/flathub/pull/1) against this branch that should fix this

Comment (tilosp) [reply com.github.micahflee.torbrowser-launcher.yaml]: Upstream PR: https://github.com/micahflee/torbrowser-launcher/pull/422

**Comment by AsciiWolf** [comment]
bot, build com.github.micahflee.torbrowser-launcher

**Comment by AsciiWolf** [comment]
bot, build com.github.micahflee.torbrowser-launcher

**Comment by AsciiWolf** [comment]
I have [found](https://github.com/micahflee/torbrowser-launcher/issues/423) one small issue, but it is nothing serious and happens on a non-Flatpak version of Tor Browser Launcher as well.

File: com.github.micahflee.torbrowser-launcher.yaml (hunk 3 above)
Comment (AsciiWolf) [reply com.github.micahflee.torbrowser-launcher.yaml]: I will probably keep the current form for now. But thanks for the tip!

 Anyway, feel free to send a PR if you want, I will happily merge it. :-)

**Comment by AsciiWolf** [comment]
@micahflee Please let us know if you are interested in write access to the repo once it is on Flathub. Thanks!

**Comment by AsciiWolf** [comment]
Upstream unfortunately didn't answer yet, but I don't have any problem with actively maintaining this Flatpak once it is on Flathub. As I said earlier, it is just the launcher, Tor Browser itself the official one without any modifications and is automatically downloaded (in a secure way) and unpacked by the launcher.

**Comment by muelli** [comment]
>  no need to update the Flatpak every time a new Tor Browser tarball is released

Are we good with non self-contained apps on flathub? Being able to start the app right away after installing from flathub rather than having to go through a shim is a nice property.

**Comment by AsciiWolf** [comment]
@muelli There are already many similar apps on Flathub, for example Albion Online. Also, Tor Browser Launcher is a preferred way to get Tor Browser in both Ubuntu and Fedora.

**Comment by muelli** [comment]
> There are already many similar apps on Flathub

fair enough. Still, being able to install and run is a nice property. But making exceptions for special cases seems reasonable. 

> for example Albion Online.

heh. yeah. And it seems to have its challenges as with maintaining state as your example shows: https://github.com/flathub/com.albiononline.AlbionOnline/issues/15


>  Also, Tor Browser Launcher is a preferred way to get Tor Browser in both Ubuntu and Fedora.

I'd say that the problem there is that the Torbrowser cannot be run in a multi-user environment and hence there is no other way for Ubuntu and Fedora. I think they'd prefer shipping the app rather than a shim.

**Comment by AsciiWolf** [comment]
> I'd say that the problem there is that the Torbrowser cannot be run in a multi-user environment and hence there is no other way for Ubuntu and Fedora. I think they'd prefer shipping the app rather than a shim.

Yep, and there is no other way (without using ugly hacks like copying Tor Browser profile to /tmp) to ship Tor Browser in a Flatpak as well. Install and run would be nice, but it would bring many issues.

**Comment by rugk** [comment]
Any news here? Can't we just merge this?

The maintainer apparently does not respond anyway, and if they later do, we could always add them. :smiley: 

Also the test command from above ends with a 404 error (are these test builds deleted after some days/weeks?)
```sh
$ flatpak install --user https://dl.flathub.org/build-repo/7084/com.github.micahflee.torbrowser-launcher.flatpakref
Fehler: Can't load uri https://dl.flathub.org/build-repo/7084/com.github.micahflee.torbrowser-launcher.flatpakref: Server returned status 404: Not Found
```

**Comment by LeandroStanger** [comment]
bot, build com.github.micahflee.torbrowser-launcher

File: com.github.micahflee.torbrowser-launcher.yaml [hunk 4]
```diff
@@ -0,0 +1,128 @@
+app-id: com.github.micahflee.torbrowser-launcher
+runtime: org.kde.Platform
+runtime-version: '5.12'
```
Comment (LeandroStanger) [inline com.github.micahflee.torbrowser-launcher.yaml]: Update runtime to 5.13

**Comment by AsciiWolf** [comment]
bot, build com.github.micahflee.torbrowser-launcher

**Comment by micahflee** [comment]
Hey, I'm upstream. Sorry about not responding! Sure I'd be happy to have write access to the Flatpak repo. But also @AsciiWolf I'd be happy if you continue to maintain it.

**Comment by rugk** [comment]
I guess you can be both maintainers. (I hope :wink:)

**Comment by AsciiWolf** [comment]
@bilelmoussaoui Feel free to remove the _awaiting-upstream_ label. :-)

**Comment by barthalion** [comment]
Looks good. I'm not happy it's not a full application but until we have extra-data checker properly implemented in CI and some kind of GPG verification in f-b, that's the best what we have. At least it doesn't escape the sandbox.

/merge

**Comment by barthalion** [comment]
@micahflee Make sure to accept the invitation to collaborators to get write access to the repository above.

**Pull request closed without merging**